### ISSUE TITLE
Rendering RealmProvider inside Expo's <Slot/> causes infinite loop

### ISSUE THREAD

**Issue by tekyu** [realm/realm-js#6823]
### How frequently does the bug occur?

Always

### Description

I'm following Expo's [documentation on Authentication](https://docs.expo.dev/router/reference/authentication/) and want to render `<RealmProvider/>` only when the user has been authenticated. 
Right now my app has the following structure

- app
   - (app)
      - _layout.tsx
      - index.tsx
   - _layout.tsx
   
Where `app/_layout.tsx` consists of 
   ```
   <AppProvider id={config.appId}>
      <UserProvider>
        <RealmProvider>
          <Slot />
        </RealmProvider>
      </UserProvider>
    </AppProvider>
   ```
   
   `app/(app)/_layout.tsx`
   ```
    <Slot />
   ```
   
   and `app/(app)/index.tsx`
   ```
   <Text>Index</Text>
   ```
   
   It works fine but as soon as I move the `<RealmProvider />` to app/(app)/_layout.tsx to have an auth stack (different pages for login/signup etc)
```
<RealmProvider>
  <Slot />
</RealmProvider>
```
so `app/_layout.tsx` looks like this
```
<AppProvider id={config.appId}>
   <UserProvider fallback={<Redirect href="/auth2" />}>
      <Slot />
   </UserProvider>
</AppProvider>
```

It just starts to re-render infinitely. Am I missing something very obvious as to why you can't put `<RealmProvider />` in a `<Slot />` or `<Stack/>` and render it down the app? It genuinely feels like one of the easiest flows to think of (rendering Realm only in protected routes) and I can't wrap my head around why it might not work.

If it's really a bug rather than my misunderstanding of either Expo's router or usage of Realm I'm including minimal reproducible code in the reproduction steps 

### Stacktrace & log output

```shell
Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
```

### Can you reproduce the bug?

Always

### Reproduction Steps

- app
   - (app)
      - _layout.tsx
      - index.tsx
   - _layout.tsx

`app/_layout.tsx`
   ```
<Slot />
   ```

`app/(app)/_layout.tsx`
```
<RealmProvider>
   <Slot />
</RealmProvider>
```
`app/(app)/index.tsx`
```
<Text>Test</Text>
```
which loops on `app/(app)/_layout.tsx` file

### Version

react-native@0.74.1

### What services are you using?

Both Atlas Device Sync and Atlas App Services

### Are you using encryption?

No

### Platform OS and version(s)

EAS/Android Studio Emulator

### Build environment

Which debugger for React Native: ..

**Comment by nirinchev** [comment]
Since you can consistently repro this and it sounds like this is a fairly minimal project, would you mind sharing it so that we can run it and investigate more easily?

**Comment by nirinchev** [comment]
Also, can you share the versions of realm js and realm react that you're using?

**Comment by tekyu** [comment]
Sure, repo is here https://github.com/tekyu/realmprovider-slot
I'm using realm@12.12.1 and @realm/react@0.9.0

I've found however that it's not `<Slot/>` that is causing this problem but wrapping it with `<AppProvider>`. When I remove it, the whole app works without looping but as soon as I wrap everything in the provider it's breaking

**Comment by wennals** [comment]
I'm having the same issue. I'm using expo with navigation and this is causing an infinite loop when I wrap my Tabs in the RealmProvider

**Comment by wennals** [comment]
@tekyu  I was able to find a work around since I'm using nested tabs. I wanted to wrap my whole app using the Tab components, but instead I wrapped the child stack and it works. 
Moved from here:
<img width="353" alt="Screenshot 2024-08-08 at 4 16 24 AM" src="https://github.com/user-attachments/assets/609fafe1-5148-4963-bac6-2a81ba97fe54">
<img width="746" alt="Screenshot 2024-08-08 at 4 16 43 AM" src="https://github.com/user-attachments/assets/41cdcfcf-0024-4272-9764-2eb9bb0ce8c9">

To here (root stack)
<img width="751" alt="Screenshot 2024-08-08 at 4 10 42 AM" src="https://github.com/user-attachments/assets/b8c8f6db-ce49-4930-88f4-b4cc60fe336c">

**Comment by isuckatlinux** [comment]
Hello, I am facing a similar issue, if I wrap the native stack navigator all my componentes will be rendering indefinetly.

`Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
    in Unknown
`
`<RealmProvider>
    <Stack.Navigator initialRouteName="MedicamentoListScreen">
        <Stack.Screen name="MedicamentoListScreen" component={MedicamentoListScreen} />
        <Stack.Screen name="MedicamentoScreen" component={MedicamentoScreen} />
    </Stack.Navigator>
</RealmProvider>`

**Comment by tekyu** [comment]
@wennals yes, I've included in the issue description that putting it in the root layout works, it does not work however if you put RealmProvider in any navigation stack
